File: NinePSharp/messages/Rremove.cs
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Rremove tag[2]
public readonly struct Rremove : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Rremove;
    public ushort Tag { get; }

    public Rremove(ushort tag)
    {
        Tag = tag;
        Size = NinePConstants.HeaderSize;
    }

    public Rremove(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
    }

    public void WriteTo(Span<byte> data)
    {
        data.WriteHeaders(Size, Tag, Type);
    }
}

File: NinePSharp.Messages.FSharp/messages/Rremove.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Rremove =
    { Size: uint32
      Tag: uint16 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Rremove
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Rremove span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        { Size = size; Tag = tag }

